### ISSUE TITLE
Schedule Private Lessons

### ISSUE THREAD

**Issue by NerdsLogic** [openstudioproject/openstudio#294]
It would be useful for dance studios, or Martial Arts studios for example, to simplify the task of scheduling a private lesson.  A private lesson will be between one or more students.

In order to schedule a private lesson right now, the process is tedious and should be simplified.  A private lesson isn't much different than a group class as a record except that it doesn't really need "spaces" "online spaces" "booking spaces"  It should show up in the location calendar (eventualy location/room) to indicate the location/room will be in use.

Some studios will have multiple lessons in the same space or room, so overlap detection is better than overlap prevention.  (perhaps a nag or confirmation screen when booking overlaps?  Option to forbid overlaps?)

**Comment by edwinvandeven** [comment]
Yes it's a feature I've been planning for a while. I would say that there might be another layer of complexity required compared to regular classes. 

To allow online booking
- Multiple teachers should be able to provide availability for a time slot
- We should be able to specify what kinds of privates are available
- We should be able to specify which teacher is able to give which kind of private classes
- Private classes should be able to have a different pricing structure then regular classes, perhaps also having special cards only for privates, to allow for package deals.

I agree with your point about overlap detection, it allows for greater flexibility. 
It might make most sense to rewrite the table structure, so events, classes and privates can all be stored in the same table. This allows for easier detection of overlapping items using quick and simple queries. Though it would take much longer to implement.

**Comment by NerdsLogic** [comment]
As to "Private classes should be able to have a different pricing structure then regular classes, perhaps also having special cards only for privates, to allow for package deals."

As far as pricing and scheduling of private classes, it feels to me like private classes are not much different than group classes except that from the system perspective a private class tends to have 1 student and could demand exclusive use of the space the lesson is conducted in.  For that matter, when the system is aware of multiple rooms per location, some users may wish to prevent overlaps as a default or an option for some or all classes OR as on option for some or all spaces (room, studio, booth, bldg, suite)

For private classes I just created class cards that have private lesson pricing.  I created a Class Card Group called Privates and allowed booking via that class card group in the private class.  For one off classes a user would pay the drop in rate which is set to a single private lesson rate.

To me what IS unique about private lessons are:

- that for self service scheduling, you have already realized the complexity.

- That the details of the private lesson would show up on a location schedule but for PUBLIC, the booking appears as "PRIVATE CLASS" instead of revealing details about the instructor and the student associated with the class for privacy purposes.  The details of the private class could be available to employees / teachers and not customers nor public when seen in the Schedule for each location, for example

- that they are tedious to set up considering that many private lessons are scheduled as one offs.  It could be handled nicely with a Class Template type of page.  An existing class in the schedule already is, of sorts, a template except that I can't use a class as template to simplify the creation of another one that has many of it's same parameters because I can't located the Duplicate link described in the quickstart manual.

- Private lessons are often booked by the instructor between the instructor and the student, so the system being private lesson friendly doesn't require it to handle student executed scheduling options at this time.  The system can be massaged to handle privates better and self service scheduling could be an enhancement at a later time.
